profiling/src/pcntl.rs

```diff
@@ -2,90 +2,124 @@ use crate::bindings::{
     datadog_php_install_handler, datadog_php_zif_handler, zend_execute_data, zend_long, zval,
     InternalFunctionHandler,
 };
-use crate::{RequestLocals, PROFILER, REQUEST_LOCALS};
+use crate::{Profiler, PROFILER, REQUEST_LOCALS};
 use log::{error, warn};
-use std::cell::RefMut;
 use std::ffi::CStr;
 use std::mem::{forget, swap};
+use std::sync::atomic::Ordering;
 
 static mut PCNTL_FORK_HANDLER: InternalFunctionHandler = None;
 static mut PCNTL_RFORK_HANDLER: InternalFunctionHandler = None;
 static mut PCNTL_FORKX_HANDLER: InternalFunctionHandler = None;
 
-fn handle_child_fork(mut locals: RefMut<RequestLocals>, retval: &mut zval) {
-    let result: Result<zend_long, _> = retval.try_into();
-    match result {
-        Ok(pid) => {
-            // The child gets pid of 0. For now, stop profiling for safety.
-            if pid == 0 {
-                match PROFILER.lock() {
-                    Ok(mut profiler) => {
-                        let mut new_profiler = None;
-                        swap(&mut *profiler, &mut new_profiler);
-
-                        /* Due to the swap, this has the old profiler.
-                         * Forget about it, it has garbage in it.
-                         */
-                        forget(new_profiler);
-                        locals.profiling_enabled = false;
-
-                        /* When we fully support forking remember:
-                         *  - Reset last_cpu_time and last_wall_time
-                         *  - Reset process_id and runtime-id tags.
-                         */
-                    }
-                    Err(err) => {
-                        error!(
-                            "Forked child failed to acquire profiler lock; a crash is likely: {}",
-                            err
-                        )
+fn handle_pcntl_fork(
+    name: &str,
+    handler: unsafe extern "C" fn(*mut zend_execute_data, *mut zval),
+    execute_data: *mut zend_execute_data,
+    return_value: *mut zval,
+) {
+    /* Hold mutexes across the handler. If there are any spurious wakeups by
+     * the threads while the fork is occurring, they cannot acquire locks
+     * since this thread holds them, preventing a deadlock situation.
+     */
+    let mut profiler_lock = PROFILER.lock().unwrap();
+    if let Some(profiler) = profiler_lock.as_ref() {
+        profiler.fork_prepare();
+    }
+
+    // Safety: we're calling the original handler with correct args.
+    unsafe { handler(execute_data, return_value) };
```

Comment:
We are chaining handlers together, seems good. I did not see this in previous design, could this be an issue in previous version ?

Reply:
We were doing the chaining outside of this helper function before, see https://github.com/DataDog/dd-trace-php/pull/1752/files#diff-3e3ad200a04052b3bf8073c003fa246b44ec22899b71d08b721acf18c5d878a5L80. I moved it inside so we could hold things across the fork easier.